# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [101]:
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline
#data="C
#:/Users/meteu/OneDrive/Desktop/SPRINGBOARD/Unit 6/1589505302_05152020_updated_ski_data_-_updated_ski_data.csv"

In [45]:
os.getcwd()

'C:\\Users\\meteu\\OneDrive\\Desktop\\SPRINGBOARD\\Unit 6\\data'

In [4]:
path="C:/Users/meteu/OneDrive/Desktop/SPRINGBOARD/Unit 6/data"
print(os.chdir(path))

None


**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [135]:

df=pd.read_csv('step3_output.csv')
df.head()

Index(['Unnamed: 0', 'Name', 'state', 'summit_elev', 'vertical_drop', 'trams',
       'fastEight', 'fastSixes', 'fastQuads', 'quad', 'triple', 'double',
       'surface', 'total_chairs', 'Runs', 'TerrainParks', 'LongestRun_mi',
       'SkiableTerrain_ac', 'Snow Making_ac', 'daysOpenLastYear', 'yearsOpen',
       'averageSnowfall', 'AdultWeekday', 'AdultWeekend', 'projectedDaysOpen',
       'NightSkiing_ac', 'clusters'],
      dtype='object')

In [103]:
df=df.drop(columns=['Unnamed: 0'])
df.head()

Name       state  summit_elev  vertical_drop  \
0               Hilltop Ski Area      Alaska         2090            294   
1            Sunrise Park Resort     Arizona        11100           1800   
2  Yosemite Ski & Snowboard Area  California         7800            600   
3               Donner Ski Ranch  California         8012            750   
4                  June Mountain  California        10090           2590   

   trams  fastEight  fastSixes  fastQuads  quad  triple  ...  \
0      0        0.0          0          0     0       1  ...   
1      0        0.0          0          1     2       3  ...   
2      0        0.0          0          0     0       1  ...   
3      0        0.0          0          0     0       1  ...   
4      0        0.0          0          2     0       0  ...   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0               30.0       30.000000        150.000000       36.0   
1              800.0       80.000000        115.000000       49.0   
2               88.0      174.873239        110.000000       84.0   
3              505.0       60.000000        163.000000       82.0   
4             1500.0      174.873239        115.103943       58.0   

   averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0             69.0     30.000000      34.00000              152.0   
1            250.0     74.000000      78.00000              104.0   
2            300.0     47.000000      47.00000              107.0   
3            400.0     75.000000      75.00000              170.0   
4            250.0     57.916957      64.16681              128.0   

   NightSkiing_ac  clusters  
0            30.0         0  
1            80.0         1  
2             0.0         1  
3             0.0         1  
4             0.0         1  

[5 rows x 26 columns]

In [104]:
#Let us see the column name
df.columns

Index(['Name', 'state', 'summit_elev', 'vertical_drop', 'trams', 'fastEight',
       'fastSixes', 'fastQuads', 'quad', 'triple', 'double', 'surface',
       'total_chairs', 'Runs', 'TerrainParks', 'LongestRun_mi',
       'SkiableTerrain_ac', 'Snow Making_ac', 'daysOpenLastYear', 'yearsOpen',
       'averageSnowfall', 'AdultWeekday', 'AdultWeekend', 'projectedDaysOpen',
       'NightSkiing_ac', 'clusters'],
      dtype='object')

In [105]:
#Let us check the data type of each columns
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 167 entries, 0 to 166
Data columns (total 26 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Name               167 non-null    object 
 1   state              167 non-null    object 
 2   summit_elev        167 non-null    int64  
 3   vertical_drop      167 non-null    int64  
 4   trams              167 non-null    int64  
 5   fastEight          167 non-null    float64
 6   fastSixes          167 non-null    int64  
 7   fastQuads          167 non-null    int64  
 8   quad               167 non-null    int64  
 9   triple             167 non-null    int64  
 10  double             167 non-null    int64  
 11  surface            167 non-null    int64  
 12  total_chairs       167 non-null    int64  
 13  Runs               167 non-null    float64
 14  TerrainParks       167 non-null    float64
 15  LongestRun_mi      167 non-null    float64
 16  SkiableTerrain_ac  167 non

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [106]:
dfo=df.select_dtypes(include=['object']) # select object type columns

df1 = pd.concat([df.drop(dfo, axis=1), pd.get_dummies(dfo)], axis=1)
df1.columns

#df=pd.concat([df.drop(columns=['state']), pd.get_dummies(df['state'])], axis=1)
df1.head()

summit_elev  vertical_drop  trams  fastEight  fastSixes  fastQuads  quad  \
0         2090            294      0        0.0          0          0     0   
1        11100           1800      0        0.0          0          1     2   
2         7800            600      0        0.0          0          0     0   
3         8012            750      0        0.0          0          0     0   
4        10090           2590      0        0.0          0          2     0   

   triple  double  surface  ...  state_Rhode Island  state_South Dakota  \
0       1       0        2  ...                   0                   0   
1       3       1        0  ...                   0                   0   
2       1       3        1  ...                   0                   0   
3       1       5        2  ...                   0                   0   
4       0       4        1  ...                   0                   0   

   state_Tennessee  state_Utah  state_Vermont  state_Virginia  \
0                0           0              0               0   
1                0           0              0               0   
2                0           0              0               0   
3                0           0              0               0   
4                0           0              0               0   

   state_Washington  state_West Virginia  state_Wisconsin  state_Wyoming  
0                 0                    0                0              0  
1                 0                    0                0              0  
2                 0                    0                0              0  
3                 0                    0                0              0  
4                 0                    0                0              0  

[5 rows x 226 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [107]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X= df1.drop(['AdultWeekend'], axis=1)
# Declare a response variable, called y, and assign it the AdultWeekend column of the df1 
y= df1['AdultWeekend']

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X)

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [108]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [109]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [110]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred=model.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [111]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model

In [112]:
from sklearn.metrics import explained_variance_score
from sklearn.metrics import mean_absolute_error
explained_variance_score(y_test, y_pred)

0.43536052089533117

In [113]:
mean_absolute_error(y_test, y_pred)

8.4285836694942

**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [114]:
model.intercept_

56.98154969939835

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [115]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
coef=pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])
coef=coef.sort_values(by=['Coefficient'],ascending=False)
coef.head(15)

Coefficient
AdultWeekday                          3.815276
trams                                 3.672430
Runs                                  2.752813
fastSixes                             2.486697
Name_Enchanted Forest Ski Area        2.145508
state_New Hampshire                   2.073624
state_Vermont                         2.023327
state_Oregon                          1.952475
Name_Apple Mountain                   1.809090
Name_Hyland Ski & Snowboard Area      1.802267
state_Wisconsin                       1.787545
Name_Kelly Canyon Ski Area            1.741937
vertical_drop                         1.700022
Name_McCauley Mountain Ski Center     1.578470
Name_New Hermon Mountain              1.540275

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [116]:
df2=df.drop(columns=['state'], axis=1)

dfo2=df2.select_dtypes(include=['object']) # select object type columns

df2 = pd.concat([df2.drop(dfo2, axis=1), pd.get_dummies(dfo2)], axis=1)
df2.columns

Index(['summit_elev', 'vertical_drop', 'trams', 'fastEight', 'fastSixes',
       'fastQuads', 'quad', 'triple', 'double', 'surface',
       ...
       'Name_White Pine Ski Area', 'Name_Whitecap Mountain',
       'Name_Wildcat Mountain', 'Name_Winterplace Ski Resort', 'Name_Wisp',
       'Name_Wolf Ridge Ski Resort', 'Name_Woodbury Ski Area',
       'Name_Woods Valley Ski Area', 'Name_Yawgoo Valley',
       'Name_Yosemite Ski & Snowboard Area'],
      dtype='object', length=191)

In [117]:
# Declare an explanatory variable, called X2,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df2
X2= df2.drop(['AdultWeekend'], axis=1)
# Declare a response variable, called y2, and assign it the AdultWeekend column of the df2 
y2= df2['AdultWeekend']

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X2 
scaler = preprocessing.StandardScaler().fit(X2)

# Declare a variable called X_scaled3, and assign it the result of calling the transform() method with parameter X3 
X_scaled2=scaler.transform(X2)

In [118]:
# Get the 1-dimensional flattened array of our response variable y2 by calling the ravel() function on y2
y2 = y2.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled2 and y2 
# Declare four variables, X_train2, X_test2, y_train2 and y_test2 separated by commas 
X_train2, X_test2, y_train2, y_test2 = train_test_split(X_scaled2, y2, test_size=0.25, random_state=1)

In [126]:
lm2 = linear_model.LinearRegression()
model2 = lm2.fit(X_train2,y_train2)

In [127]:
# Make a variable called y_pred2 and assign it the result of calling predict() on our model variable with parameter X_test2
y_pred2=model.predict(X_test2)
explained_variance_score(y_test2, y_pred2)

0.4559989952335791

In [128]:
mean_absolute_error(y_test2, y_pred2)

7.910737536700698

In [129]:
model2.intercept_

56.167011029796065

In [130]:
coef2=pd.DataFrame(abs(lm2.coef_), X2.columns, columns=['Coefficient'])
coef2=coef2.sort_values(by=['Coefficient'],ascending=False)
coef2.head(15)

Coefficient
AdultWeekday                         3.817529
Name_Enchanted Forest Ski Area       2.126470
Name_Suicide Six                     1.970589
vertical_drop                        1.537331
Name_Gunstock                        1.479195
Name_Kelly Canyon Ski Area           1.427547
Runs                                 1.404886
Name_Monarch Mountain                1.285855
fastQuads                            1.184552
clusters                             1.173088
Name_New Hermon Mountain             1.132249
Name_Bromley Mountain                1.125837
Name_Christmas Mountain              1.122013
Name_Hyland Ski & Snowboard Area     1.119571
Name_Magic Mountain Ski Area         1.079908

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [136]:
df3=df.drop(columns=['state', 'summit_elev'], axis=1)

dfo3=df3.select_dtypes(include=['object']) # select object type columns

df3 = pd.concat([df3.drop(dfo3, axis=1), pd.get_dummies(dfo3)], axis=1)
df3.columns

Index(['Unnamed: 0', 'vertical_drop', 'trams', 'fastEight', 'fastSixes',
       'fastQuads', 'quad', 'triple', 'double', 'surface',
       ...
       'Name_White Pine Ski Area', 'Name_Whitecap Mountain',
       'Name_Wildcat Mountain', 'Name_Winterplace Ski Resort', 'Name_Wisp',
       'Name_Wolf Ridge Ski Resort', 'Name_Woodbury Ski Area',
       'Name_Woods Valley Ski Area', 'Name_Yawgoo Valley',
       'Name_Yosemite Ski & Snowboard Area'],
      dtype='object', length=191)

In [137]:
# Declare an explanatory variable, called X3,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df3
X3= df3.drop(['AdultWeekend'], axis=1)
# Declare a response variable, called y3, and assign it the AdultWeekend column of the df3
y3= df3['AdultWeekend']

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X3 
scaler = preprocessing.StandardScaler().fit(X3)

# Declare a variable called X_scaled3, and assign it the result of calling the transform() method with parameter X3 
X_scaled3=scaler.transform(X3)

In [138]:
# Get the 1-dimensional flattened array of our response variable y3 by calling the ravel() function on y3
y3 = y3.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled3 and y3 
# Declare four variables, X_train3, X_test3, y_train3 and y_test3 separated by commas 
X_train3, X_test3, y_train3, y_test3 = train_test_split(X_scaled3, y3, test_size=0.25, random_state=1)

In [139]:
lm3 = linear_model.LinearRegression()
model3 = lm3.fit(X_train3,y_train3)

In [140]:
# Make a variable called y_pred3 and assign it the result of calling predict() on our model variable with parameter X_test3
y_pred3=model.predict(X_test3)
explained_variance_score(y_test3, y_pred3)

0.4637519215852912

In [141]:
mean_absolute_error(y_test3, y_pred3)

7.886242598140238

In [142]:
model3.intercept_

56.10929544631756

In [143]:
coef3=pd.DataFrame(abs(lm3.coef_), X3.columns, columns=['Coefficient'])
coef3=coef3.sort_values(by=['Coefficient'],ascending=False)
coef3.head(15)

Coefficient
AdultWeekday                         3.830467
Name_Enchanted Forest Ski Area       2.111256
Name_Suicide Six                     1.936411
Name_Gunstock                        1.483028
vertical_drop                        1.425496
Runs                                 1.417882
Name_Monarch Mountain                1.378831
Name_Kelly Canyon Ski Area           1.358789
fastQuads                            1.282764
clusters                             1.149282
Name_Ski Granby Ranch                1.117429
Name_Big Snow Resort - Blackjack     1.103983
Name_Christmas Mountain              1.091551
Name_Bromley Mountain                1.083639
Name_Hyland Ski & Snowboard Area     1.077611

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. |0.435 | 8.428 | |
| Model 2. | 0.456|7.910 |'state'|
| Model 3. | 0.463 |7.886 |'state', 'summit_elev'|

Model Selection:

**<font color='teal'> The best model is Model 3 because it does have a highest Explained Variance and the lowest Mean Absolute Error. </font>**